### DISCUSSION TITLE
How do you configure your app? Not a bundle, not an environment, but your actual application?

### DISCUSSION THREAD

**Discussion by jerrac** [symfony/symfony#47960] [Q&A] [answered]
I feel like I'm missing something obvious. But how in the world do you add configuration for your application?

Specifically in Symfony 6.1.

I've found docs on configuration for bundles/packages and docs on how to configure some of those settings via environment variables. I found the section on services.yaml's `parameters` section for reusable values in config. I've watched some SymfonyCasts on Bundles. But I have yet to find anything on either the docs site, or via web searches, on how to add normal application config.

As a super simple example, how would I make the site name configurable?

If I was starting from scratch, I'd just set a `$config['site_name'] = foobar;` line in my config.php and be done. I thought Symfony would have something like that. So, somewhere in the config directory I could create a custom `.yaml` file for application specific config, and then I'd be able to find that config via Symfony's config class.

Something like: `config/app.yaml`
```yaml
app:
  site_name: foobar
```
But `php bin/console config:dump app` just throws an `No extensions with configuration available for "app".` error.

I thought maybe a DependencyInjection folder, Configuraton class, and AppExtension class in the main `src` folder would do something, but it doesn't. Assuming I'm not screwing that up, which I probably am since my attempts at bundle config haven't been very successful either...

Anyway, any help is appreciated. I'm pretty stuck. :\

**Comment by ghost** [comment, marked as answer]
You have to register extension in `Kernel.php` by calling `$container->registerExtension()` inside `Kernel::configureContainer()`

**Reply by jerrac** [reply]
Sorry, could you expand on your answer? Or link me to where in the docs it's explained? 

I can't find anything in the code for a `Kernel::configureContainer()` method. PHPStorm doesn't list that as one of the methods you can override with their generate menu. And I didn't see anything in the `Symfony\Component\HttpKernel\Kernel` code either.

I'm assuming that registering my `AppExtension` class is what you mean I need to do, and that doing so will enable my configuration to be found. Right?

**Reply by ghost** [reply]
Ok, I see that in 6.1 everything is hidden inside MicroKernelTrait. You have to declare method `configureContainer` as protected in Kernel.php (copy past what is already in this method in this trait) and inside call `$container->registerExtension()`.

Btw. if you provided some code snippets it would be easier to reason about your problem.

**Reply by jerrac** [reply]
Ok, that gave me the hints I needed to take how bundles work and translate that into something for my app.

When I load the default Symfony page, my config is dumped.

Thanks!

My code:

`config/app/app.yaml`:
```yaml
app:
  site_name: "Awesome"
```

In `Kernel.php`:
```php
	protected function configureContainer(
      ContainerConfigurator $container,
      LoaderInterface $loader,
      ContainerBuilder $builder
    ): void {
        $configDir = $this->getConfigDir();

        $container->import($configDir.'/{packages}/*.{php,yaml}');
        $container->import($configDir.'/{packages}/'.$this->environment.'/*.{php,yaml}');

        if (is_file($configDir.'/services.yaml')) {
            $container->import($configDir.'/services.yaml');
            $container->import($configDir.'/{services}_'.$this->environment.'.yaml');
        } else {
            $container->import($configDir.'/{services}.php');
        }
        $builder->registerExtension(new AppExtension());

        if (is_dir($configDir.'/app/')) {
            $container->import($configDir.'/app/*.{yaml,php}');
        }
    }
```

`src/DependencyInjection/AppExtension.php`:
```php
<?php

namespace App\DependencyInjection;

use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Extension\ExtensionInterface;

class AppExtension extends \Symfony\Component\HttpKernel\DependencyInjection\Extension implements ExtensionInterface
{

    /**
     * @inheritDoc
     */
    public function load(array $configs, ContainerBuilder $container)
    {
        var_dump($configs);
        die(1);
    }


}
```

`src/DependencyInjection/Configuration.php`:
```php
<?php

namespace App\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;

class Configuration implements \Symfony\Component\Config\Definition\ConfigurationInterface
{

    /**
     * @inheritDoc
     */
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder('app');
        $rootNode = $treeBuilder->getRootNode();
        $rootNode->children()->scalarNode('site_name')->defaultValue('App')->end()->end();

        return $treeBuilder;
    }

}
```
